### PR TITLE
Add typescript friendly mixins

### PR DESCRIPTION
**Pull request by jrobinson01**
adds typescript ports of the routing mixins:
routing.mixin.ts and animated-routing.mixin.ts
This PR also includes the built .js versions of those files. These _might_ play nicer with projects that are not using typescript, but are using tsc and jsdoc. This PR should not affect projects using closure compiler.

### CODE REVIEW

**Review by chrisgubbels — APPROVED** [review]

File: lib/animated-routing.mixin.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import {RoutingMixin, type Constructor, type RoutingMixinInterface} from './routing.mixin.js';
+
+// type Constructor<T = {}> = new (...args: any[]) => T;
+
+// export declare class RoutingMixinInterface {
+//   routeEnter(currentNode: RouteTreeNode, nextNodeIfExists: RouteTreeNode | undefined, routeId: string, context: Context): Promise<boolean | void>;
+//   routeExit(currentNode: RouteTreeNode, nextNode: RouteTreeNode | undefined, routeId: string, context: Context): Promise<boolean | void>;
+// }
+
+function animatedRoutingMixin<T extends Constructor<LitElement>>(Superclass:T, className:string) {
```
Comment (barronhagerman) [inline lib/animated-routing.mixin.ts:13]: Should this be specific to `LitElement`? 
```suggestion
function animatedRoutingMixin<T extends Constructor<HTMLElement>>(Superclass:T, className:string) {
```

File: lib/routing.mixin.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import type RouteData from './route-data.js';
+import type RouteTreeNode from './route-tree-node.js';
+import type { LitElement } from 'lit';
+
+export type Constructor<T = {}> = new (...args: any[]) => T;
+export declare class RoutingMixinInterface {
+  routeEnter(currentNode: RouteTreeNode, nextNodeIfExists: RouteTreeNode | undefined, routeId: string, context: Context): Promise<boolean | void>;
+  routeExit(currentNode: RouteTreeNode, nextNode: RouteTreeNode | undefined, routeId: string, context: Context): Promise<boolean | void>;
+}
+const RoutingMixin = <T extends Constructor<LitElement>>(superclass: T) => {
```
Comment (barronhagerman) [inline lib/routing.mixin.ts:13]: Same here. Should this be specific to `LitElement`? 
```suggestion
const RoutingMixin = <T extends Constructor<HTMLElement>>(superclass: T) => {
```

Comment (matt-sneed) [reply lib/routing.mixin.ts:11]: This was something we discovered in origin. If we don't use LitElement as the super class we get linter errors when using the lifecycle methods like connectedCallback etc

https://lit.dev/docs/composition/mixins/#mixins-in-typescript

File: lib/animated-routing.mixin.ts (hunk 1 above)
Comment (matt-sneed) [reply lib/animated-routing.mixin.ts:5]: i commented below but without the LitElement as the extended class, it causes typescript errors when using the lifecycle methods. @barronhagerman @jrobinson01

**Pull request merged**